src/Cassandra/Requests/RequestExecution.cs

```diff
@@ -501,47 +501,66 @@ private void PrepareAndRetry(byte[] id, Host host)
                     .ContinueSync(_ =>
                     {
-                        Send(request, host, ReprepareResponseHandler);
+                        Send(request, host, NewReprepareResponseHandler(ex));
                         return true;
                     });
                 return;
             }
-            Send(request, host, ReprepareResponseHandler);
+            Send(request, host, NewReprepareResponseHandler(ex));
         }
 
         /// <summary>
         /// Handles the response of a (re)prepare request and retries to execute on the same connection
         /// </summary>
-        private void ReprepareResponseHandler(IRequestError error, Response response, Host host)
+        private Action<IRequestError, Response, Host> NewReprepareResponseHandler(
+            PreparedQueryNotFoundException originalError)
         {
-            try
+            void ResponseHandler(IRequestError error, Response response, Host host)
             {
-                if (error?.Exception != null)
-                {
-                    HandleRequestError(error, host);
-                    return;
-                }
-                RequestExecution.ValidateResult(response);
-                var output = ((ResultResponse)response).Output;
-                if (!(output is OutputPrepared outputPrepared))
+                try
                 {
-                    _parent.SetCompleted(
-                        new DriverInternalError("Expected prepared response, obtained " + output.GetType().FullName));
-                    return;
-                }
+                    if (error?.Exception != null)
+                    {
+                        HandleRequestError(error, host);
+                        return;
+                    }
+
+                    RequestExecution.ValidateResult(response);
+                    var output = ((ResultResponse) response).Output;
+                    if (!(output is OutputPrepared outputPrepared))
+                    {
+                        _parent.SetCompleted(new DriverInternalError("Expected prepared response, obtained " + output.GetType().FullName));
+                        return;
+                    }
+
+                    if (!outputPrepared.QueryId.SequenceEqual(originalError.UnknownId))
+                    {
+                        throw new PreparedStatementIdMismatchException(
+                            originalError.UnknownId,
+                            "ID mismatch while trying to reprepare (expected " 
+                            + $"{BitConverter.ToString(originalError.UnknownId).Replace("-", "")}, " 
+                            + $"got {BitConverter.ToString(outputPrepared.QueryId).Replace("-", "")}). " 
+                            + "This prepared statement won't work anymore. " 
+                            + "This usually happens when you run a 'USE...' query after " 
+                            + "the statement was prepared.");
```

Comment:
NIT: The message building could be done within the `PreparedStatementIdMismatchException`, passing the `originalError.UnknownId` and `originalError.UnknownId` as parameters.

Reply:
+1 on that, I'll push that change before merging